sdk/python/feast/infra/online_stores/contrib/cassandra_online_store/cassandra_online_store.py

```
@@ -477,98 +482,73 @@ def on_failure(exc, concurrent_queue):
+                        ):
+                            feature_value = timestamp
+                        else:
+                            feast_value_type = valProto.WhichOneof("val")
+                            if feast_value_type is None:
+                                feature_value = None
+                            elif feast_value_type in feast_array_types:
+                                feature_value = getattr(
+                                    valProto, str(feast_value_type)
+                                ).val
+                            else:
+                                feature_value = getattr(valProto, str(feast_value_type))
+                        feature_values += (feature_value,)
+
+                    feature_values = feature_values + (
+                        entity_key_bin,
+                        timestamp,
+                    )
+                    insert_cql = self._get_cql_statement(
+                        config,
+                        "insert_sorted_features",
+                        fqtable=fqtable,
+                        ttl=int(ttl),
+                        session=session,
+                        feature_names_str=feature_names_str,
+                        params_str=params_str,
+                    )
+                    batch.add(insert_cql, feature_values)
+                    batch_count += 1
+
+                    if (
+                        online_store_config.write_batch_size is None
+                        or batch_count >= online_store_config.write_batch_size
+                    ):
+                        CassandraOnlineStore._apply_batch(
+                            rate_limiter,
+                            batch,
+                            progress,
+                            session,
+                            concurrent_queue,
+                            on_success,
+                            on_failure,
                         )
-                        if ttl >= 0:
```

This should not be removed as its not just a validation for user provided TTL value. 
So there are 2 ways of setting TTL based on the value of `apply_ttl_on_write` configuration:
1. If the `apply_ttl_on_write` is set to True, then we will use the user provided ttl directly (in the insert statement), in which case it does not matter if you remove the above `if ttl >= 0` statement and do the validation instead at feature view level.
2.  But if the `apply_ttl_on_write` is set to False (which means TTL will be set based on event timestamp vs the write based ttl where its is calculated as now()+TTL as in 1. ), then the calculated ttl can be negative if a row is already expired. For example, lets say if the ttl is set to 5 min by the user, but we are materializing the data from yesterday, then in this case the data is already expired, in which case we don't need to ingest this data/add this row to the batch. This is why I'm previously checking the ttl even before processing each row to add it to the batch.

---

Regular fvs only use the online store config ttl value so a negative ttl would cause all inserts to be skipped. Maybe we should raise an error in that case?

---

Oh yeah thats right. The cassandra server throws an error. But we don't need to raise an error because its already happening here or below that:
https://github.com/ExpediaGroup/feast/blob/252a9cfb5cb34a1bf08497fc37932e156d0a51fb/sdk/python/feast/infra/online_stores/contrib/cassandra_online_store/cassandra_online_store.py#L616
All the errors during batch insert are already being thrown.

---

`_get_ttl()` is not handling Feature View TTL as 0. 0 means never expire.

---

That happens only if online_store_config.key_ttl_seconds is set to 0. Refer  https://github.com/ExpediaGroup/feast/blob/764d5a4113e1920f0cc365b67e9de9949cb7e248/sdk/python/feast/infra/online_stores/contrib/cassandra_online_store/cassandra_online_store.py#L1019-L1023